src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs

```diff
@@ -336,17 +336,38 @@ private HttpMessageHandler SetupHandlerChain()
             CheckDisposed();
             HttpMessageHandler handler = _handler ?? SetupHandlerChain();
 
-            Exception error = ValidateAndNormalizeRequest(request);
+            Exception error = ValidateAndNormalizeRequest(request, out var shouldBufferContent);
             if (error != null)
             {
                 return Task.FromException<HttpResponseMessage>(error);
             }
 
+            if (shouldBufferContent)
+            {
+                return SendWithBufferedContentAsync(handler, request, cancellationToken);
+            }
+
             return handler.SendAsync(request, cancellationToken);
         }
 
-        private Exception ValidateAndNormalizeRequest(HttpRequestMessage request)
+        private async Task<HttpResponseMessage> SendWithBufferedContentAsync(
+            HttpMessageHandler handler, HttpRequestMessage request, CancellationToken cancellationToken)
         {
+            await request.Content.LoadIntoBufferAsync(HttpContent.MaxBufferSize, cancellationToken).ConfigureAwait(false);
```

Comment:
I'm weary that we're going to buffer up to 2GB here. I'd almost prefer a smaller maximum with fallback to non-chunked if we're over. Thoughts @dotnet/ncl ?

Reply:
I suppose we could add a property to control the max buffer size, but that introduces new API.

I don't really like having a fallback -- it can lead to unexpected behavior. Either it should buffer or fail (if there's a limit).

Reply:
Agreed. Predictability is important.